src/libos/src/process/do_futex.rs

```diff
@@ -144,16 +144,15 @@ pub async fn futex_wait_bitset(
     // Must make sure that no locks are holded by this thread before wait
     drop(futex_bucket);
 
-    waiter.wait(/*timeout.as_ref()*/).await;
-
-    /* TODO: handle errror
-    if let Err(e) = res {
+    let mut timeout = timeout.map_or(None, |t| Some(t));
+    // Wait until we wake it up or reach timeout.
+    if let Err(_) = waiter.wait_timeout(timeout.as_mut()).await {
         let (_, futex_bucket_ref) = FUTEX_BUCKETS.get_bucket(futex_item.key);
         let mut futex_bucket = futex_bucket_ref.lock().unwrap();
         futex_bucket.dequeue_item(&futex_item);
-        return_errno!(e.errno(), "futex wait timeout or interrupted");
```

Comment:
Actually, I think you can keep the original code of returning `e.errno()` as it would save you from changing it again when `Waiter::wait_timeout` can be interrupted.

Reply:
But there is a problem: we return `ETIME` in `wait_timeout`, but here we need to return `ETIMEDOUT`.
Should we return `ETIMEDOUT` in `wait_timeout`?

Reply:
ETIME  Timer expired (POSIX.1 (XSI STREAMS option)).

              (POSIX.1 says "STREAM ioctl(2) timeout".)

       ETIMEDOUT
              Connection timed out (POSIX.1-2001).

Reply:
I don't even know there is an errno code named `ETIME`! Could you give an example system call that returns `ETIME`?

Still, my judgement (you can challenge) favors `ETIMEDOUT`. The reason is that from the user perspective, he/she calls `Waiter::wait`, not create a timer directly. The `Waiter::wait` method is most like futex wait, which would return `ETIMEDOUT` if timeout.